### PR TITLE
V0.8.0 - Avro support

### PR DESCRIPTION
**Pull request by pjm17971**
The goal of this change is to enable serialization of TimeSeries using [Avro](https://avro.apache.org/).

Avro is a way of taking a data structure (say, the TimeSeries, Event, or TimeRange currently), along with a schema known to both ends of the transfer, and generating an efficient and type safe buffer to send.

**Event changes**

In order to achieve this we enable true subclassing of Events, including cleaning up the current Event class hierarchy (or lack thereof). `Event` is now a base class and subclasses are now: `TimeEvent`, `TimeRangeEvent` and `IndexedEvent`. These can be further subclassed to provided schema or for other reasons.

 **Event** (base class with static Event methods)
    |-----**TimeEvent** (the old Event)
    |-----**TimeRangeEvent**
    |-----**IndexedEvent**

[Note: `TimeEvent` replaces `Event` for non-static functionality, which is likely to be a problematic breaking change, but I think is clearer going forward]

Here's an example. First define a subclass, specifying a static function for the `dataSchema` of your event:

```
class TrafficEvent extends TimeEvent {
    constructor(arg1, arg2) {
        super(arg1, arg2);
    }

    static dataSchema() {
        return {
            type: "record",
            fields: [
                {name: "name", type: "string"},
                {name: "in", type: "long"},
                {name: "out", type: "long"}
            ]
        };
    }
}
```

Now you can create an instance of that event type. That event can then be outputted to Avro and sent across the wire:

```
const event = new TrafficEvent(1445449170000, {name: "source1", in: 2, out: 11});
const buffer = event.toAvro(); // send me
```

On the receiving end a new Event can be created from the buffer:

```
const buffer = ... // received buffer 
const event = new TrafficEvent(buffer);
```

**TimeSeries support**

In addition these subclassed events can be used within a `TimeSeries` or `Collection` and the resulting series can also be serialized out to avro. To do this we need to subclass the TimeSeries in order to define the schema for the meta data of the TimeSeries as well as the Event type the TimeSeries should use:

```
class StatusSeries extends TimeSeries {
    constructor(arg) {
        super(arg);
    }

    metaSchema() {
        return [{name: "name", type: "string"}];
    }

    static event(key) {
        return StatusEvent;
    }
}
```

Now use the new TimeSeries class, and serialize it out to avro:

```
const timeseries = new StatusSeries({
    name: "traffic",
    columns: ["time", "value", "status"],
    points: [
        [1400425947000, 52, "ok"],
        [1400425948000, 18, "ok"],
        [1400425949000, 26, "fail"],
        [1400425950000, 93, "offline"]
    ]
 });

const buffer = timeseries1.toAvro();  // send me
```

As for the event serialization, the receiving end of the buffer can create a new TimeSeries from the buffer:

```
const timeseries = new StatusSeries(buffer);
```

### CODE REVIEW

**Comment by jdugan1024** [comment]
@pjm17971 Do you know why the tests are failing in CI?

**Comment by pjm17971** [comment]
@jdugan1024 Yeah, it's because the performance tests are running slower there

**Comment by pjm17971** [comment]
Also note, the merge here is into v0.8, so don't worry about doc strings etc. I need to do a complete pass over that before merging into master and releasing. Mostly this is to make it easier for monte to track changes.

**Comment by montegoode** [comment]
@pjm17971 Many thanks sir.

File: src/pond/lib/base/util.js [hunk 1]
```diff
@@ -244,5 +245,65 @@ export default {
...
+
+    //
+    // Functions to turn constructor args
+    // into other stuff
+    //
+
+    timestampFromArg(arg) {
+        if (_.isNumber(arg)) {
+            return new Date(arg);
+        } else if (_.isString(arg)) {
```
Comment (pjm17971) [inline src/pond/lib/base/util.js]:
```
> const d = new Date("2016-09-17")
undefined
> d
2016-09-17T00:00:00.000Z
> 
```

**Comment by pjm17971** [comment]
Yep, basically this PR is a conversation about how to do this.

Yes, you need to define a schema for your TimeSeries or Event, via sub-classing. Basically this is saying that if you want to talk in structured data, you need to define the type and that type's schema ahead of time, and essentially share that type between the two ends of the conversation. I'd imagine in a real system (between python and JS) that you'd factor out the schema parts and move them to a common import, but I've not really thought about that too much. Between JS you could `import { TrafficEvent } from esnet-pond-types"` for example and just use the JS classes to build your data on either side of the conversation.

JSON serialization has no schema other than the first level of subclassing (TimeEvent, IndexedEvent etc), so I don't see any necessity to change the way that works with generic types. But yeah, I guess it becomes a poor man's serialization. 

I think deep data should work, although I haven't tried it. You just need to define your Event's schema to have records within records or arrays within the records, and it should serialize if the schema matches the data.

File: src/pond/lib/base/util.js (hunk 1 above)
Comment (pjm17971) [inline src/pond/lib/base/util.js]: Oh, I see what you mean. I'm not sure that was the original intention here.

**Pull request merged**